src/us/kbase/workspace/test/database/provenance/ProvenanceActionTest.java

```diff
@@ -0,0 +1,1059 @@
+	}
+	
+	private void failWithResolvedObjects(
+			final ProvenanceAction.Builder b,
+			final List<String> input,
+			final String exception) {
+		try {
+			b.withResolvedObjects(input);
+			fail("expected exception");
+		} catch (Exception got) {
+			TestCommon.assertExceptionCorrect(got, new IllegalArgumentException(exception));
+		}
+	}
+	
+	@Test
+	public void withIncomingArgsFail() throws Exception {
+		failWithIncomingArgs(
+				list("foo", null, "baz"),
+				"Null or whitespace only string in collection incoming args");
+		failWithIncomingArgs(
+				list("foo", "baz", "   \t    "),
+				"Null or whitespace only string in collection incoming args");
+	}
+	
+	private void failWithIncomingArgs(final List<String> input, final String exception) {
+		try {
+			ProvenanceAction.getBuilder().withIncomingArgs(input);
+			fail("expected exception");
+		} catch (Exception got) {
+			TestCommon.assertExceptionCorrect(got, new IllegalArgumentException(exception));
+		}
+	}
+	
+	@Test
+	public void withOutgoingArgsFail() throws Exception {
+		failWithOutgoingArgs(
+				list("foo", null, "baz"),
+				"Null or whitespace only string in collection outgoing args");
+		failWithOutgoingArgs(
+				list("foo", "baz", "   \t    "),
+				"Null or whitespace only string in collection outgoing args");
+	}
+	
+	private void failWithOutgoingArgs(final List<String> input, final String exception) {
+		try {
+			ProvenanceAction.getBuilder().withOutgoingArgs(input);
+			fail("expected exception");
+		} catch (Exception got) {
+			TestCommon.assertExceptionCorrect(got, new IllegalArgumentException(exception));
+		}
+	}
+	
+	@Test
+	public void withExternalDataFail() throws Exception {
+		try {
+			ProvenanceAction.getBuilder().withExternalData(list(
+					ExternalData.getBuilder().withDataID("d").build(),
+					null
+					));
+			fail("expected exception");
+		} catch (Exception got) {
+			TestCommon.assertExceptionCorrect(got, new NullPointerException(
+					"Null item in external data"));
+		}
+	}
+	
+	@Test
+	public void withSubActionsFail() throws Exception {
+		try {
+			ProvenanceAction.getBuilder().withSubActions(list(
+					SubAction.getBuilder().withName("a").build(),
+					null
+					));
+			fail("expected exception");
+		} catch (Exception got) {
+			TestCommon.assertExceptionCorrect(got, new NullPointerException(
+					"Null item in subactions"));
+		}
+	}
+	
+	@Test
+	public void withCustomFail() throws Exception {
+		final Map<String, String> nullMap = new HashMap<>();
+		nullMap.put("a", "b");
+		nullMap.put(null, "c");
+		nullMap.put("d", "e");
+		try {
+			ProvenanceAction.getBuilder().withCustom(nullMap);
+			fail("expected exception");
+		} catch (Exception got) {
+			TestCommon.assertExceptionCorrect(got, new IllegalArgumentException(
+					"Null key in custom provenance"));
+		}
+	}
+	
+	@Test
+	public void buildFail() throws Exception {
+		try {
+			ProvenanceAction.getBuilder().build();
```

Comment:
Have you also tested the case where there are fields provided, but they're all null?

Reply:
Yes. Yes I did